src/fsharp/TypeChecker.fs

```diff
@@ -5572,14 +5574,16 @@ and TcExprUndelayed cenv overallTy env tpenv (expr: SynExpr) =
           TcStmtThatCantBeCtorBody cenv env tpenv e1
 
     | SynExpr.IfThenElse (e1,e2,e3opt,spIfToThen,isRecovery,mIfToThen,m) ->
-        let e1',tpenv = TcExprThatCantBeCtorBody cenv cenv.g.bool_ty env tpenv e1  
-        (if isNone e3opt && not isRecovery then UnifyTypes cenv env m overallTy cenv.g.unit_ty)
```

Comment:
You should not change the order in which this "unit" type constraint is applied in the case of ``if/then``.  So you need to keep this line.  I think you can specify ``{ env with specializedErrorMessage = Some FSComp.SR.missingElseBranch } `` on this line?